File: Solutions/LeetCode/0336-PalindromePairs/Solution.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution {
    private bool IsPalindrome(string s, int left, int right) {
        while (left < right) {
            if (s[left++] != s[right--]) {
                return false;
            }
        }
        return true;
    }

    public IList<IList<int>> PalindromePairs(string[] words) {
        var res = new List<IList<int>>();
        var wordToIdx = new Dictionary<string, int>();
        for (int i = 0; i < words.Length; i++) {
            wordToIdx[words[i]] = i;
        }

        for (int i = 0; i < words.Length; i++) {
            string word = words[i];
            int n = word.Length;
            for (int j = 0; j <= n; j++) {
                // Case 1: Prefix is a palindrome, check if reversed suffix is in dictionary
                if (IsPalindrome(word, 0, j - 1)) {
                    char[] suffixArr = word.Substring(j).ToCharArray();
                    Array.Reverse(suffixArr);
                    string revSuff = new string(suffixArr);
                    if (wordToIdx.TryGetValue(revSuff, out int idx) && idx != i) {
                        res.Add(new List<int> { idx, i });
                    }
                }

                // Case 2: Suffix is a palindrome, check if reversed prefix is in dictionary
                // j < n to avoid duplicates (suffix must be non-empty)
                if (j < n && IsPalindrome(word, j, n - 1)) {
                    char[] prefixArr = word.Substring(0, j).ToCharArray();
                    Array.Reverse(prefixArr);
                    string revPref = new string(prefixArr);
                    if (wordToIdx.TryGetValue(revPref, out int idx) && idx != i) {
                        res.Add(new List<int> { i, idx });
                    }
                }
            }
        }

        return res;
    }

    public static void Main(string[] args) {
        Solution sol = new Solution();
        var testCases = new (string[] words, int[][] expected)[] {
            (new string[] { "abcd", "dcba", "lls", "s", "sssll" }, new int[][] { new int[] { 0, 1 }, new int[] { 1, 0 }, new int[] { 3, 2 }, new int[] { 2, 4 } }),
            (new string[] { "bat", "tab", "cat" }, new int[][] { new int[] { 0, 1 }, new int[] { 1, 0 } }),
            (new string[] { "a", "" }, new int[][] { new int[] { 0, 1 }, new int[] { 1, 0 } })
        };

        for (int idx = 0; idx < testCases.Length; idx++) {
            var tc = testCases[idx];
            var result = sol.PalindromePairs(tc.words);
            
            // Compare results
            bool pass = true;
            if (result.Count != tc.expected.Length) {
                pass = false;
            } else {
                var resSorted = result.Select(r => string.Join(",", r.OrderBy(x => x))).OrderBy(x => x).ToList();
                var expSorted = tc.expected.Select(e => string.Join(",", e.OrderBy(x => x))).OrderBy(x => x).ToList();
                for (int i = 0; i < resSorted.Count; i++) {
                    if (resSorted[i] != expSorted[i]) {
                        pass = false;
                        break;
                    }
                }
            }

            Console.WriteLine($"Test {idx + 1}: words=[{string.Join(", ", tc.words)}] -> ResultCount={result.Count} | Pass={pass}");
        }
    }
}

File: Solutions/LeetCode/0336-PalindromePairs/Solution.py
class Solution:
    def palindromePairs(self, words: list[str]) -> list[list[int]]:
        word_to_idx = {w: i for i, w in enumerate(words)}
        res = []
        
        for i, word in enumerate(words):
            n = len(word)
            for j in range(n + 1):
                pref = word[:j]
                suff = word[j:]
                if pref == pref[::-1]:
                    rev_suff = suff[::-1]
                    if rev_suff in word_to_idx and word_to_idx[rev_suff] != i:
                        res.append([word_to_idx[rev_suff], i])
                if suff != "" and suff == suff[::-1]:
                    rev_pref = pref[::-1]
                    if rev_pref in word_to_idx and word_to_idx[rev_pref] != i:
                        res.append([i, word_to_idx[rev_pref]])
        return res

if __name__ == '__main__':
    sol = Solution()
    test_cases = [
        (["abcd", "dcba", "lls", "s", "sssll"], [[0, 1], [1, 0], [3, 2], [2, 4]]),
        (["bat", "tab", "cat"], [[0, 1], [1, 0]]),
        (["a", ""], [[0, 1], [1, 0]])
    ]
    for idx, (words, expected) in enumerate(test_cases, 1):
        result = sol.palindromePairs(words)
        print(f"Test {idx}: words={words} -> Result={result} | Expected={expected} | {'PASS' if sorted(result) == sorted(expected) else 'FAIL'}")
